core/src/main/scala-3/neotypes/CypherStringInterpolator.scala

```diff
@@ -0,0 +1,129 @@
+            // 2. the hashtag at the end of string: "foo bar#"
+            //    It is unusual that cypher query ends with `#`
+            //    so, optimistically use `part.init` here.
+            Left(
+              part.init
+            ) // Left means "nth part is `part.init` and the interpolation after the nth part is a plain value"
+          else
+            // Right means "nth part is`part` and the interpolation after the nth part (if any) is a query parameter"
+            Right(part)
+        }
+        .toList
+    }
+    val interleaved: Expr[List[Res]] =
+      '{
+        Interleave($partsPacked, $stringFragmentsPartitionedByInterpolationKind)
+      }
+
+    '{
+      createQuery($interleaved: _*)
+    }
+}
+
+object Interleave {
+  import CypherStringInterpolator.Res
+
+  /** @param interpolations
+    *   a list of interpolation generator functions. If a function receives `true` it should generate QueryArg
+    *   representation and otherwise it should generate String representation.
+    * @param stringFragments
+    *   a list of string splitted by interpolation and distinguished by interpolation kind. If the n-th element of
+    *   `stringFragments` is `Right`, it implies the next adjacent interpolation should be a `QueryArg` interpolation.
+    *   If the n-th element of `stringFragments` is `Left`, it implies the next adjacent interpolation should be a plain
+    *   value interpolation.
+    */
+  def apply(interpolations: List[Boolean => Res], stringFlagments: List[Either[String, String]]): List[Res] =
+    interleave(interpolations, stringFlagments, List.empty)
+  private[this] def interleave(
+    interpolations: List[Boolean => Res],
+    stringFlagments: List[Either[String, String]],
+    acc: List[Res]
+  ): List[Res] =
+    (stringFlagments, interpolations) match
+      case (Nil, Nil) => acc
+      // this should be always query interpolations because if there is a plain value interpolation, there should be `#` prefix.
+      case (Nil, interps) => acc.reverse ++ (interps.map(_.apply(true)))
+      case (Left(nthPartExpectPlainValueInsert) :: tail, interp :: interps) =>
+        interleave(interps, tail, interp(false) :: Left(nthPartExpectPlainValueInsert) :: acc)
+      case (Right(nthPartExpectQueryValueInsert) :: tail, interp :: interps) =>
+        interleave(interps, tail, interp(true) :: Left(nthPartExpectQueryValueInsert) :: acc)
+      case (Right(theLastPart) :: Nil, Nil) => (Left(theLastPart) :: acc).reverse
+      case (Left(theLastPartIsAHashTag) :: Nil, Nil) =>
+        (Left(theLastPartIsAHashTag) :: acc).reverse
+      case (Right(partExpectQueryValueInsert) :: _, Nil) =>
+        // this should be unreachable
+        throw new Exception(s"interpolation value number mismatch")
+      case (Left(partExpectPlainValueInsert) :: _, Nil) =>
+        // this should be unreachable
+        throw new Exception(s"interpolation value number mismatch")
```

Comment:
Maybe just use `case _ =>` here to catch both cases.
Also, couldn't we use here the same `abortMacro` thing?

Reply:
`report` APIs are in `quotes.reflect` package and  you need `Quotes` in scope, so I think it is not appropriate here.

Reply:
Okay, but then how does this exception interact with the macro? Or will this get thrown during runtime?
It doesn't matter since it is impossible, I just want to understand better how this works :)

Reply:
>  I just want to understand better how this works :)

When you throw an exception inside `inline def aMethodThatUsesQuoteAPI`, the exception is thrown at macro expansion.

For example, if I put a line that throw an exception(see `macroImpl` method),

```scala
final case class CypherStringInterpolator(private val sc: StringContext) extends AnyVal {
  inline def c(inline args: Any*): DeferredQueryBuilder = ${ CypherStringInterpolator.macroImpl('args, 'sc) }
}

object CypherStringInterpolator {

  type Res = Either[String, QueryArg]

  def macroImpl(argsExpr: Expr[Seq[Any]], clz: Expr[StringContext])(using quotes: Quotes): Expr[DeferredQueryBuilder] =
    throw new Exception("boom") // <--- exception here
    import quotes.reflect.*

    def invert[T: Type](exprs: Expr[Seq[T]]): List[Expr[T]] =
      exprs match {
        case Varargs(props)                => props.toList
        case '{ Seq(${ Varargs(props) }) } => props.view.map(_.asExprOf[T]).toList
        case _                             => report.errorAndAbort(s"Unable to match ${exprs.show}")
      }

```
And then I run `++3.3.0 core/test:compile`
  I got the following.

```
sbt:root> ++3.3.0 core/test:compile
[info] Setting Scala version to 3.3.0 on 2 projects.
[info] Excluded 14 projects, run ++ 3.3.0 -v for more details.
[info] Reapplying settings...
[info] set current project to root (in build file:/path/to/neotypes/)
[warn] sbt 0.13 shell syntax is deprecated; use slash syntax instead: core / Test / compile
[info] compiling 1 Scala source to /path/to/neotypes/core/target/scala-3.3.0/test-classes ...
[error] -- Error: /path/to/neotypes/core/src/test/scala/neotypes/syntax/CypherQueryInterpolationSpec.scala:12:29 
[error] 12 |    val (query, params, _) = c"CREATE (a: Test { name: 'John' })".build()
[error]    |Exception occurred while executing macro expansion.
[error]    |java.lang.Exception: boom
[error]    |    at neotypes.query.CypherStringInterpolator$.macroImpl(CypherStringInterpolator.scala:17)
[error]    |
```

On the other hand, when I throw an exception outside a quote scope, the exception is thrown at runtime.

For example

```scala
  def apply(interpolations: List[Boolean => Res], stringFlagments: List[Either[String, String]]): List[Res] =
    throw new Exception("boom") <--- exception here
    interleave(interpolations, stringFlagments, List.empty)
```

and `++3.3.0 core/test:compile` is quiet, but `++3.3.0 core/test` throws the exception at runtime.